### PR TITLE
[mystrom] Add support for myStrom Bulb

### LINKED ISSUE #9818: [mystrom] contribution, add support for myStrom bulb
The myStrom Binding currently supports only the Plug. I would like to implement support for the Bulb.

### PR DESCRIPTION
**Pull request by Fredo70**
Add support for myStrom Bulb
Add properties to myStrom devices and an action to refresh the properties.
closes #9818 

Signed-off-by: Frederic Chastagnol <fchastagnol@fredoware.ch>

### CODE REVIEW

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:112]: see above

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:120]: check SAT report and remove unnecessary empty lines.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:158]: You can make that a `private static final` field

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:176]: see above

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:189]: see above, also check other occurences

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:213]: Maybe surround with `try-catch` for `IllegalArgumentException`, otherwise the scheduler might crash

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:251]:
```suggestion
    private String getDeviceSpecificInformation() throws MyStromException {
        String url = getHostname() + "/api/v1/device";
        try {
            Request request = getHttpClient().newRequest(url).timeout(10, TimeUnit.SECONDS).method(HttpMethod.GET);
            ContentResponse response = request.send();
            if (response.getStatus() != HTTP_OK_CODE) {
                throw new MyStromException(
                    "Error sending HTTP GET request to " + url + ". Got response code: " + response.getStatus());
            }
            return response.getContentAsString();   
        } catch (InterruptedException | TimeoutException | ExecutionException e) {
            throw new MyStromException(COMMUNICATION_ERROR + e.getMessage());
        }
    }
```

Also in other occurences. Usually it makes code more readable to keep it to gether and not separate it with exception handling code.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:284]: What's wrong with using the `FormContentProvider`? I would prefer that ove rthe hard-to-understand-code below. I

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:51]: Maybe rename to `MyStromPlugHandler` to keep the class naming consistent.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:87]: If you make `gson` in the abstract handler `protected final` instead of `private final` you can access the field directly and remove the getter.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:115]: Did you think about moving all these methods (`sendHttpGet` and `sendHttpPost` and one more I can't remember) to the abstract handler? If another device is added, it'll likely use one of these and the code can be shared.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandlerFactory.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandlerFactory.java:46]:
```suggestion
    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_PLUG, THING_TYPE_BULB);
```

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandlerFactory.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandlerFactory.java:68]:
```suggestion
        } else if (THING_TYPE_BULB.equals(thingTypeUID)) {
```

File: bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml:57]:
```suggestion
			<channel id="brightness" typeId="system.brightness"/>
```

File: bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml:113]:
```suggestion
		<description>Transition time from the light’s current state to the new state.</description>
```

Unit not needed, the item has the unit.

File: bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml:134]: `readonly=false` is the default, not needed.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:236]: If you move the `sendHttpGet` to the abstract handler, you can re-use the code and completely remove this method.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:170]: Change the fields to `protected`, then you don't need the getters

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:71]: We usually add `Abstract` at the beginning, so `AbstractMyStromHandler` would be a better naming.

**Review by J-N-K — CHANGES_REQUESTED** [review]
Thanks for your first contribution. I have left some comments inside. If you need assistance, plese ping me.

File: bundles/org.openhab.binding.mystrom/README.md
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/README.md:66]: Sorry. I don't understand, what you mean with "de done on a regular basis". I asked for advice in the community: https://community.openhab.org/t/is-this-best-practice-use-an-action-to-refresh-properties/114140

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:16]: I think yes. I use intellij and setup the environment as described in the documentation (inclusive the eclipse plugin). I run "spotless:apply" before compiling.

File: bundles/org.openhab.binding.mystrom/README.md
Comment (J-N-K) [reply bundles/org.openhab.binding.mystrom/README.md:66]: I think updating during initialize would be the better way. The user can trigger that by disabling/enabling the thing.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:96]: Ok. Why not. I was just wondering that most other addons use "Collections.singleton".

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:100]: Without `@Nullable` I get a warning in intellij "_Expression 'getDeviceInfo()' might evaluate to null but is assigned to a variable that is annotated with @NotNull_". So I thought, I have to insert this.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:141]: Same as above. If I annotate the method with `@Nullable` intellij says "_@Nullable method 'getDeviceInfo' always returns a non-null value_".

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:100]: That’s an IntelliJ issue. IntelliJ does not properly support Eclipse Null-annotations.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:120]: The generated html files in target directory? They show neither errors nor warnings.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:284]: Ok. I will test again. I meant I couldn't get it to work with that, because the hsv String is separated with a semicolon ;

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:51]: Ok. I wasn't sure if I could rename the class as it was there before.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:87]: I learned this as a best practice a long time ago ;-) I can do it like this... it's also easier.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandlerFactory.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandlerFactory.java:68]: Is that `else` not superfluous. I think this is very controversial on the net. I don't care. I can insert `else`.

File: bundles/org.openhab.binding.mystrom/README.md
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/README.md:66]: Ok. It is already updated during initialize. I will remove the action.

Comment (J-N-K) [reply bundles/org.openhab.binding.mystrom/README.md:66]: Maybe you can add a note to the documentation instead thatdisabling/enabling the thing can be used to update the properties. Not sure if that is clear to everybody without reading the code.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:51]: That's not a problem because it's not exposed to the user and also an internal package.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:115]: Good. I wasn't sure about that either. In the original addon there was only one Thing and the method was already called sendHttpGet. But the GET requests that support the plug are only supported by the plug, for example. On the other hand, the Plug does not support POST requests from the Bulb. So these methods are device-specific. Perhaps the name is inconveniently chosen and they should be private.

Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:115]: Ah, with the getDeviceSpecificInformation method below, that does make sense.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:87]: Sorry for my inexperience with Github. Can I adapt the changes you suggested directly in my code and push a new commit? Or do I have to commit them here?

Comment (J-N-K) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromAbstractHandler.java:87]: If you press "commit suggestion" your branch is updated witrh the change. You have to `git pull` to get this changes in your local copy. Usually if it's only a minor change and no other changes are needed, it's ok to use that feature. If other changes are needed, in most cases applying the changes to you local copy and push them all together is easier.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:284]: Test again. Doesn't work. The `FormContentProvider` does `encode(...)`. If it's encoded, it doesn't work.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:120]: Running install issues the following warning:
```
Jan. 23, 2021 5:16:26 NACHM. org.openhab.tools.analysis.checkstyle.KarafFeatureCheck processFiltered
WARNUNG: Could not find file feature file features/karaf/openhab-addons/src/main/feature/feature.xml
```
I have no idea what it is and how and whether I have to correct anything.

Comment (J-N-K) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:120]: You can ignore that. It's a bug in the static analysis tool.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:284]: https://github.com/eclipse/jetty.project/blob/1701b83fd5035cbd9ab88dd7f3b210190776a6c0/jetty-client/src/main/java/org/eclipse/jetty/client/util/FormContentProvider.java#L53-L63 is a better (easier to understand) implementation. And please remove the FormContentProvider(no commented code).

After that please push your changes, the PR will automatically be updated. Do not use force-push, otherwise the changes are not visible and harder to review.

**Review by J-N-K — CHANGES_REQUESTED** [review]

**Comment by Fredo70** [comment]
I think I have adjusted everything. But before I commit, I have one more question.
Is it possible, for the channel "ColorTemperature", to set the default widget automatically as a slider with min=1, max=18 and step=1.
I looked in other addons and thought it was done with the `<state min="1" max="18" step="1" pattern="%d"/>` in the channel-type. But it always just comes up as a number with the analysis and I have to set the default widget manually.

**Comment by J-N-K** [comment]
Not sure. Maybe adding `<category>slider</category>` helps.

File: bundles/org.openhab.binding.mystrom/README.md
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/README.md:33]:
```suggestion
The properties are updated during initialize.
Disabling/enabling the thing can be used to update the properties.
```

Start each sentence in the docimentation on a new line. This makes reviewing later documentation changes easier.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:82]: Move these to the bottom of the class.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:93]: Move up (under the other `private static final`), so constants are kept together. Maybe also rename to `DEVICE_INFO_MAP_TYPE` or something like that.

**Review by J-N-K — CHANGES_REQUESTED** [review]
Just some last small comments. Otherwise looks good to me.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java
Comment (Fredo70) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromBulbHandler.java:93]: Moved up. But I remove the `COMMUNICATION_ERROR` constant. It was not used since it is in the abstract class. I had not seen that.

**Review by J-N-K — APPROVED** [review]

**Comment by J-N-K** [comment]
Thanks. Waiting for Jenkins before merge.

**Comment by J-N-K** [comment]
Jenkins failed prior to building mystrom. Rebuild.

**Review by pail23 — APPROVED** [review]
LGTM

File: bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml:56]: Please use the system color temperature channel type `system.color-temperature` for this and map your values 1-18 to 0-100% for it.

File: bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.mystrom/src/main/resources/OH-INF/thing/thing-types.xml:130]: This can be removed.

**Review by kaikreuzer — CHANGES_REQUESTED** [review]

**Comment by Fredo70** [comment]
There is still DCO issue because I used once "commit suggestion". Can I do something to make the check pass?

**Comment by fwolter** [comment]
Kai's findings seem to be addressed appropriately.

**Pull request merged**
